File: core/java/src/main/java/org/aiddl/core/java/function/numerical/AdditionFunction.java
package org.aiddl.core.java.function.numerical;

import org.aiddl.core.java.interfaces.Function;
import org.aiddl.core.java.representation.NumericalTerm;
import org.aiddl.core.java.representation.Term;

public class AdditionFunction implements Function {

	@Override
	public Term apply(Term x) {
		NumericalTerm sum = Term.integer(0);
		for ( int i = 0 ; i < x.size() ; i++ ) {
			sum = sum.add(x.get(i).asNum());
		}
		return sum;
	}
}

File: core/scala/src/main/scala/org/aiddl/core/scala/function/numerical/AdditionFunction.scala
package org.aiddl.core.scala.function.numerical

import org.aiddl.core.scala.function.Function
import org.aiddl.core.scala.representation.{Num, Term, Tuple}

protected[function] class AdditionFunction extends Function {
  def apply(x: Term): Term = x match {
    case Tuple(args@_*) => args.reduce(_.asNum + _.asNum)
    case _ => throw new IllegalArgumentException(s"Bad argument: $x. Need tuple of numerical terms.")
  }
}
